Alternator: enable tablets by default - depending on tablets_mode_for_new_keyspaces

nyh opened the pull request:
Before this series, Alternator's CreateTable operation defaults to creating a table replicated with vnodes, not tablets. The reasons for this default included missing support for LWT, Materialized Views, Alternator TTL and Alternator Streams if tablets are used. But today, all of these (except the still-experimental Alternator Streams) are now fully available with tablets, so we are finally ready to switch Alternator to use tablets by default in new tables.

We will use the same configuration parameter that CQL uses, tablets_mode_for_new_keyspaces, to determine whether new keyspaces use tablets by default. If set to `enabled`, tablets are used by default on new tables. If set to `disabled`, tablets will not be used by default (i.e., vnodes will be used, as before). A third value, `enforced` is similar to `enabled` but forbids overriding the default to vnodes when creating a table.

As before, the user can set a tag during the CreateTable operation to override the default choice of tablets or vnodes (unless in `enforced` mode). This tag is now named `system:initial_tablets` - whereas before this patch it was called `experimental:initial_tablets`. The rules stay the same as with the earlier, experimental:initial_tablets tag: when supplied with a numeric value, the table will use tablets. When supplied with something else (like a string "none"), the table will use vnodes. 

Fixes https://github.com/scylladb/scylladb/issues/22463

Backport to 2025.4, it's important not to delay phasing out vnodes.

nyh commented:
This pull request is based on @ScyllaPiotr's https://github.com/scylladb/scylladb/pull/26211 with my review comments applied - including some changes to the test_scan.py and three separate patches added at the end of the series.

nyh commented:
It seems the last failure is in the dtest alternator_ttl_tests.py::test_ttl_with_load_and_decommission. The last CI failure on it was https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/8686/testReport/alternator_ttl_tests/TestAlternatorTTL/FullDtest___full_split000___test_ttl_with_load_and_decommission/  but it also failed in previous runs of Alternator with tablets so it's not a fluke (see https://github.com/scylladb/scylladb/pull/26211#issuecomment-3472134982 seen by @ScyllaPiotr).

The test itself passes, but an error message appears in the log of the fourth node in the middle of the test, which upsets the test framework:
```
INFO  2025-11-02 19:34:48,647 [shard 0:strm] raft_topology - request decommission for: deebc918-98c9-40cd-8e22-86e0f0b2c5f4
INFO  2025-11-02 19:34:48,655 [shard 0: gms] raft_topology - topology cmd rpc barrier_and_drain is called index=20
INFO  2025-11-02 19:34:48,721 [shard 0: gms] raft_topology - Got raft_topology_cmd::barrier_and_drain, version 13, current version 13
INFO  2025-11-02 19:34:48,721 [shard 0: gms] raft_topology - raft_topology_cmd::barrier_and_drain done
INFO  2025-11-02 19:34:48,721 [shard 0:strm] raft_topology - decommission: waiting for completion (request ID: fe7cc2b8-b822-11f0-0aab-439cc4f11513)
INFO  2025-11-02 19:34:48,721 [shard 1: gms] raft_topology - Got raft_topology_cmd::barrier_and_drain, version 13, current version 13
INFO  2025-11-02 19:34:48,721 [shard 1: gms] raft_topology - raft_topology_cmd::barrier_and_drain done
INFO  2025-11-02 19:34:48,722 [shard 0: gms] raft_topology - topology cmd rpc barrier_and_drain completed with status=succeeded index=20
INFO  2025-11-02 19:34:48,737 [shard 0: gms] storage_proxy - update_fence_version: new fence_version 13 is set, prev fence_version 12, pending stale writes 0
INFO  2025-11-02 19:34:48,737 [shard 1: gms] storage_proxy - update_fence_version: new fence_version 13 is set, prev fence_version 12, pending stale writes 0
INFO  2025-11-02 19:34:48,889 [shard 0: gms] raft_topology - topology cmd rpc barrier_and_drain is called index=21
INFO  2025-11-02 19:34:48,923 [shard 0: gms] raft_topology - Got raft_topology_cmd::barrier_and_drain, version 15, current version 15
INFO  2025-11-02 19:34:48,923 [shard 0: gms] raft_topology - raft_topology_cmd::barrier_and_drain done
INFO  2025-11-02 19:34:48,923 [shard 1: gms] raft_topology - Got raft_topology_cmd::barrier_and_drain, version 15, current version 15
INFO  2025-11-02 19:34:48,923 [shard 1: gms] raft_topology - raft_topology_cmd::barrier_and_drain done
INFO  2025-11-02 19:34:48,923 [shard 0: gms] raft_topology - topology cmd rpc barrier_and_drain completed with status=succeeded index=21
INFO  2025-11-02 19:34:48,941 [shard 0: gms] storage_proxy - update_fence_version: new fence_version 15 is set, prev fence_version 13, pending stale writes 0
INFO  2025-11-02 19:34:48,941 [shard 1: gms] storage_proxy - update_fence_version: new fence_version 15 is set, prev fence_version 13, pending stale writes 0
ERROR 2025-11-02 19:34:48,942 [shard 0:strm] raft_topology - Decommission failed. See earlier errors (Rolled back: Failed to drain tablets: std::runtime_error (There are nodes with tablets to drain but no candidate nodes in DC dc1. Consider adding new nodes or reducing replication factor.)). Request ID: fe7cc2b8-b822-11f0-0aab-439cc4f11513
```
https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/8686/artifact/logs-full.release.000/1762112108010_alternator_ttl_tests.py%3A%3ATestAlternatorTTL%3A%3Atest_ttl_with_load_and_decommission/node4.log

If I understand what this test is doing, the test has an Alternator table (it also uses TTL, but I don't believe the TTL has anything to do with the failure) and 4 nodes, and it repeatedly decomissions the fourth node and then creates a new fourth node.

I don't understand (maybe @yarongilor who wrote this test can help) how, if decomission failed, the test didn't fail immediately and only failed at the end when we find those error messages in the log. I also don't understand (maybe @tgrabiec can help?) what the error message "There are nodes with tablets to drain but no candidate nodes in DC dc1" means. 

I have a guess which I still need to prove: Maybe we just don't wait enough for the new node to to join before asking to decomission a node? Looking at run_decommission_then_add_node(), after adding a new node it waits for 5 seconds (!). After just 5 seconds, the loop continues and the test may try to decomission one of the nodes. Maybe those 5 seconds was not enough to finish preparing the new node, and for some reason we can't drain a node from its tablets because one of them needs to go to the new node which isn't ready to accept it for some reason?

To be honest, I don't understand this problem, and will probably need some help from the tablets team. I am guessing this problem isn't specific to Alternator or to TTL, and it's a generic tablets bug that decomission isn't handled as gracefully as we should. I think we're seeing the problem in this specific Alternator test just because this test does all thse decomissions and creations which perhaps no CQL test does in the same way. The reason why we never saw this failure before this pull request is that before this pull request, this test never ran with tablets so it didn't have the issues related to tablet migration.

tgrabiec commented:
> Clearly the test should be fixed to ignore that error in the log message.

I think we should rather try to make decommission succeed, by first adding the node (which currently follows decommission), otherwise no decommission will happen and the stress is not stressing decommission handling. 

It worked with vnodes, because they allow decommission with not enough racks.

I don't know why the error is ignored here, there is nothing in the test which could make it unreliable.

nyh commented:
> > Clearly the test should be fixed to ignore that error in the log message.
> 
> I think we should rather try to make decommission succeed, by first adding the node (which currently follows decommission), otherwise no decommission will happen and the stress is not stressing decommission handling.

I don't think the order matters.  The test has a thread which in a tight loop decommissions a node, and if that's successful, adds another node. So the decommissions and additions are interleaved - it doesn't matter which is first. Since the test starts with 4 nodes, it should be fine to do the decommission first.

In any case, as I said above, the purpose of this test is to test **Alternator TTL** while bothering it with random topology operations - it's not a goal of this test to check that how these operations work or sometimes fail. If they sometimes fail, the test will just do less of those operations and it will still do its job.

> I don't know why the error is ignored here, there is nothing in the test which could make it unreliable.

I also can't think of what exactly is the scenario that led to this failure. I'm not even sure that it's not a **real** decommission bug - or maybe it's not a bug. But in any case, this is an Alternator test - not a decommission test. It shouldn't fail because of decommission bugs and apparently - the person who wrote this test saw these problems even with vnodes, otherwise they wouldn't have caught the decommission errors and ignored them. So it's not a new regression in tablets. The only "regression" is that with tablets we print an ERROR to the log that we didn't print with vnodes, and that makes dtest fail.

nyh commented:
I created https://github.com/scylladb/scylla-dtest/pull/6377 to ignore the ERROR log messages that come with the currently-ignored decommission failure.

I still don't know if it's the "correct" fix - @tgrabiec believes the decommission will always fail with tablets so this test doesn't check anything and needs to be rewritten - and yet when I run the test myself, it doesn't always fail (I can't even reproduce this bug) - I see the decommissions succeeding. So I think the test is doing what it wanted to do, even with tablets.

tgrabiec commented:
> > > I don't think the order matters. The test has a thread which in a tight loop decommissions a node, and if that's successful, adds another node. So the decommissions and additions are interleaved - it doesn't matter which is first. Since the test starts with 4 nodes, it should be fine to do the decommission first.
> > 
> > 
> > The order matters because w start with 1 node per rack. You cannot decommission the last node in a rack with tablets.
> 
> Oh, wow. Are you saying that we have 4 nodes on 4 racks? Or 4 nodes on 3 racks? Or what? Why can't we decommission the fourth node as the first step?
> 
> In any case, I ran this test myself on my laptop, and it did NOT fail when decommissining the fourth node in a 4-node cluster. In fact for me, the test worked without failing as it did in the CI - the decommission-and-then. I don't know what to make of it. It certainly doesn't _always_ fail as you say it should.

Ah. We have 4 nodes and 4 racks, but keyspace is using RF=3. After recent changes, tablet keyspaces will expand RF=3 to a rack list. So tablets will be placed in a subset of racks. So if the node you decommission is in the rack which doesn't have tablets, it will not fail, but finish without moving anything.  

If the point of the test is to actually test what happens if you decommission tablets, then we should pick a node which has replicas. That can be done like this:

```python
rf = get_replication_options(session, "ks")
if type(rf) is list:
    node_to_remove = next(n for self.cluster.nodelist() if n.rack in rf)
else:
    node_to_remove = self.cluster.nodelist()[-1] # any node will do
```

nyh commented:
> Ah. We have 4 nodes and 4 racks\

Wow. As Isidor Rabi said about the Muon - "who ordered that?" :-)

>, but keyspace is using RF=3. After recent changes, tablet keyspaces will expand RF=3 to a rack list. So tablets will be placed in a subset of racks. So if the node you decommission is in the rack which doesn't have tablets, it will not fail, but finish without moving anything.

If the table has multiple tablets, not just one, won't they be spread over all four nodes?
If not, isn't this a problem? If a user has 4 nodes and creates one table, the table will only use 3 out of the 4 nodes?

> If the point of the test is to actually test what happens if you decommission tablets, then we should pick a node which has replicas. That can be done like this:

To be honest, I don't know what is the point of this test ;-)

Tests always assumed that if we create a cluster of 4 nodes and then create a table and run something like Alternator TTL, it will happen on all 4 nodes and removing and adding nodes will "do something" to disturb the system. If there are now some magical rules that apply to when adding or removing nodes doesn't actually change anything because the tablets are pinned to the previously existing nodes and the "racks" that they belong to, then somebody will need to rethink about all the tests that involve decommissions and whether they need to be changed. I don't think it's only specific to this test - I assume we have other tests which try to decommission nodes in parallel with doing other things.

> 
> ```python
> rf = get_replication_options(session, "ks")
> if type(rf) is list:
>     node_to_remove = next(n for self.cluster.nodelist() if n.rack in rf)
> else:
>     node_to_remove = self.cluster.nodelist()[-1] # any node will do
> ```

I'll try to see whether I can reproduce the failure seen in CI with this snippet. If I understand what you're saying correctly, the current code basically removes a random node and will sometime do nothing (and pass) and sometime choose one of the 3 nodes with data and it cannot be decommisssioned.

nyh commented:
I just realized why I can't reproduce these dtest failures locally.
The failed CI run is called "dtest with tablets". I'm guessing now that it's not normal dtest but rather dtest somehow modified to enforce tablets. Now I need to figure out why or how to do it in my manual run.

tgrabiec commented:
Pass `--tablets` to pytest command line

nyh commented:
Ok, I created a better fix for the dtest, https://github.com/scylladb/scylla-dtest/pull/6377, which does some "real" decommissions and doesn't just ignore ones that fail. Please review that PR since this is the last test failing on this P1-priority PR.

annastuchlik commented on docs/alternator/new-apis.md:
```diff
@@ -181,27 +181,30 @@ entire data center, or other data centers, in that case.
-Alternator tables default to using vnodes - not tablets.
+In this version, tablet support is almost complete, so new
+Alternator tables default to following what the global configuration flag
+[`tablets_mode_for_new_keyspaces`](../reference/configuration-parameters.rst#confval-tablets_mode_for_new_keyspaces)
```
Consider removing the backquotes for the option name to render as a link.
See https://github.com/scylladb/scylladb/pull/26211/files#r2487197039.

nyh replied:
Thanks, I'll remove those backquotes in the next version I push.

nyh commented:
Pushed a new version, rebased on new master, removed backquotes in a link (suggested in review by @annastuchlik) and that's it. The dtest fix already got in, so with luck, the CI will pass.

nyh commented:
> #### Failed Tests (1/263961):
> 
>     * [test_distributed_count_all](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/21531/testReport/junit/distributed_queries_test/TestDistributedAggregations/release___dtest_with_tablets___test_distributed_count_all) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_distributed_count_all)

That test (for distributed counting) has nothing to do with Alternator or anything else changed in this PR, and it hasn't failed in previous CI runs for this PR, so it must be a flaky test, and indeed it's a known flaky test - https://github.com/scylladb/scylladb/issues/16709 - which we thought a year ago was no longer flaky and closed that issue but I think it still is flaky - and I reopened it. 

@scylladbbot trigger-ci

nyh commented:
Hallelujah, all tests including dtests passed.
@ScyllaPiotr please review this series, which is based on your own work from https://github.com/scylladb/scylladb/pull/26211.

ScyllaPiotr commented on test/alternator/test_scan.py:
```diff
@@ -412,31 +412,44 @@ def test_scan_long_partition_tombstone_string(dynamodb, query_tombstone_page_lim
         str = 'x' * 10240
-        N = 600
+        N = KB // 10
+        print(N)
```
Debug print leftover? (Might be my mistake)

nyh replied:
I'll remove it. One of the good features of pytest is that print()s are completely benign - they don't get shown unless the test failed (or you enabled them with the "-s" option). But there's indeed no particular reason for this print so I'll remove it.

ScyllaPiotr commented on alternator/executor.cc:
```diff
@@ -109,8 +109,13 @@ extern const sstring TTL_TAG_KEY("system:ttl_attribute");
 
-// Alternator-specific table properties stored as table tags, that are not considered internal:
-//
+// The following are tags also have the "system:" prefix but are NOT used
```
~are~

nyh replied:
Fixed.

ScyllaPiotr commented on alternator/executor.cc:
```diff
@@ -109,8 +109,13 @@ extern const sstring TTL_TAG_KEY("system:ttl_attribute");
-// Alternator-specific table properties stored as table tags, that are not considered internal:
-//
+// The following are tags also have the "system:" prefix but are NOT used
+// by Alternator to store table properties - only the user every writes to
```
ever~y~

nyh replied:
Thanks. Fixed.

ScyllaPiotr commented on test/cluster/test_alternator.py:
In commit msg: _create an Alternator table**~t~**_

nyh replied:
Fixed

nyh commented on db/config.cc:
You are very right:

First, no non-Alternator test required this flag to be live-updatable because I'm the only "fanatic" who cares about super fast tests in test/alternator. Other developers seem to be completely fine with test/cluster which starts a new cluster just to set some silly option. So this option was tested by those tests - but never in live-updatable.

I could add a test in test/cqlpy to check that this flag is live-updatable also for CQL, but I was too lazy to do it now. I did look at the code, though, to make sure the live-updatableness has a chance of working - and it does. What "breaks" live-updatable parameters is saving their current value in some copy in some long-living object. But I checked and the code in CQL that checks this flag checks it directly from the configuration, without saving copies anywhere. So I wasn't worried it doesn't work.

nyh commented on test/alternator/test_scan.py:
```diff
@@ -412,31 +412,44 @@ def test_scan_long_partition_tombstone_string(dynamodb, query_tombstone_page_lim
+# grow signficantly above 1MB, depending on implementation details and the
+# number of vnodes or tablets.
+# Since the primary goal of this test was just to check that the Scan doesn't
+# return the *entire* table, below we fill the data with enough data (MB=14)
```
Right. I'll fix.

nyh commented on test/alternator/test_scan.py:
```diff
@@ -412,31 +412,44 @@ def test_scan_long_partition_tombstone_string(dynamodb, query_tombstone_page_lim
+# Since the primary goal of this test was just to check that the Scan doesn't
+# return the *entire* table, below we fill the data with enough data (MB=14)
+# that not all of it is returned in a single page, even if we have up to
+# 16 vnodes or tablets (per node).
```
But, oh, I see that yes, it was 4*2 so 8. I'll fix.

nyh commented:
I can already see CI failed on 19 different instances of https://github.com/scylladb/scylladb/issues/25084. I guess Azure is down, or our connection to it is rate-limited or I don't know what :-(
